src/agent_observer.py

```diff
@@ -0,0 +1,104 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""The agent relation observer module."""
+
+
+import logging
+import os
+from pathlib import Path
+
+import ops
+
+import service
+from charm_state import AGENT_RELATION, State
+
+logger = logging.getLogger()
+JENKINS_WORKDIR = Path("/var/snap/jenkins-agent/common/agent")
+AGENT_READY_PATH = Path(JENKINS_WORKDIR / ".ready")
+
+
+class Observer(ops.Object):
+    """The Jenkins agent relation observer."""
+
+    def __init__(
+        self,
+        charm: ops.CharmBase,
+        state: State,
+        jenkins_agent_service: service.JenkinsAgentService,
+    ):
+        """Initialize the observer and register event handlers.
+
+        Args:
+            charm: The parent charm to attach the observer to.
+            state: The charm state.
+            jenkins_agent_service: Service manager that controls Jenkins agent service.
+        """
+        super().__init__(charm, "agent-observer")
+        self.charm = charm
+        self.state = state
+        self.jenkins_agent_service = jenkins_agent_service
+        charm.framework.observe(
+            charm.on[AGENT_RELATION].relation_joined, self._on_agent_relation_joined
+        )
+        charm.framework.observe(
+            charm.on[AGENT_RELATION].relation_changed, self._on_agent_relation_changed
+        )
+        charm.framework.observe(
+            charm.on[AGENT_RELATION].relation_departed, self._on_agent_relation_departed
+        )
+
+    def _on_agent_relation_joined(self, event: ops.RelationJoinedEvent) -> None:
+        """Handle agent relation joined event.
+
+        Args:
+            event: The event fired when an agent has joined the relation.
+        """
+        self.charm.unit.status = ops.MaintenanceStatus(
+            f"Setting up '{event.relation.name}' relation."
+        )
+
+        relation_data = self.state.agent_meta.get_jenkins_agent_v0_interface_dict()
+        logger.debug("Agent relation data set: %s", relation_data)
+        event.relation.data[self.charm.unit].update(relation_data)
+
+    def _on_agent_relation_changed(self, event: ops.RelationChangedEvent) -> None:
+        """Handle agent relation changed event.
+
+        Args:
+            event: The event fired when the agent relation data has changed.
+        """
+        logger.info("%s relation changed.", event.relation.name)
+
+        # Check if the pebble service has started and set agent ready.
+        if os.path.exists(str(AGENT_READY_PATH)) and self.jenkins_agent_service.is_active:
```

Comment:
What if the secret has changed?

Reply:
I think if the secret has changed, it means that the node was deleted either manually or due to an error on Jenkins's side. In that case, the agent service won't be able to connect to Jenkins and so it will be in an error state. I think this behavior is correct since in that situation an admin would need to manually resolve the root cause on the Jenkins server and reestablish the agent relation.

Reply:
Hmm, it seems valid for Jenkins to be allowed to change the secret and expect the agent to be re-added using the new secret. E.g., if the secret is leaked it might need to be changed.

Speaking of, I can't remember whether juju secrets are being used for the secret here?

Reply:
I looked into this in detail and I can confirm that even on the jenkins side there is no way of changing the secret without removing the node and recreating a new one, the secret is basically generated once when the node is created. And once the node is deleted the connection between jenkins and the agent is broken immediately and will cause the service to go into an error state.

I'd say to "rotate" this secret we'd probably need to remove the current relation and recreate it so that a new secret is generated.